Refactor CBPS functions as classes and add an example +  package it

ivanhigueram opened the pull request:
Main changes in this PR: 
 - Refactor `cbps_torch.py` as a class with the same functionality as the original.
 - Test the class using the `lalonde.R` dataset from [LalRUtils](https://github.com/apoorvalal/LalRUtils/blob/master/data/lalonde.psid.RData).
 - Create `project.toml` for future packaging and testing
 - Add some small docstrings just to guide myself. 

Some changes in the `CBPS` implementation:
 - Add L1-norm regularization to loss during optimization. Not sure if that's a desirable feature, but is there.

ivanhigueram committed:
Strucutre library and move some code to src

This commit adds a small template to build the library with code
in ./src. The cbps code is reduced to just a small class that can
be refactored to other applications or training mechanisms.

Some things to change are:
 - Figure out regularization. Added a not so fancy L1. Can we
   change the loss?
 - Apply other trainers.

ivanhigueram committed:
Test pytorch implementation using Lalonde

This commmit test the CBPS class with PyTorch with a simple example
using the Lalonde dataset.

Some things to work on:
 - regularization is in its infancy
 - numpy operations are buggy in weights calculation
 - implement some pytorch good practices when it comes to device.

ivanhigueram committed:
Add sys trick to run notebook in examples

Just add the sys.path trick to export libraries from the src/ folder
without moving the jupyter nobook to main.

apoorvalal commented on src/cbps_torch.py:
```diff
@@ -0,0 +1,112 @@
+            loss = self.loss_function(self.theta, self.X_t, self.w_t)
+
+            if self.reg:
+                loss += self.reg * torch.linalg.vector_norm(self.theta, 1)
```
nice! why prefer L1 over L2 though? Do you want sparsity?

ivanhigueram replied:
Yeah, I thought sparsity would be a good goal, but we can keep it flexible and add L2 too.

apoorvalal commented on src/cbps_retarget.py:
```diff
@@ -0,0 +1,60 @@
+    # Iteratively apply gradient descent to minimize the loss function
+    for iteration in range(niter):
+        optimizer.zero_grad()  # Clear previous gradients
+        loss = loss_function(theta, X0, X1)  # Calculate loss
```
for parity with disaggregated data case, also add regularization here. These are implicitly the same program except we only feed in summary statistics in the retarget version

ivanhigueram replied:
Huh, i was confused how to integrate those `X1`, can you expand what `retargeting` means here?

apoorvalal replied:
`retargeting` here means you're estimating balancing weights that calibrate a sample `X0` towards a population `X1`, where you only have summary statistics for the target population `X1`. This is useful in a bunch of settings where external validity is something you try to achieve rather than handwave :laughing:.  
In the ATT estimation case, you have individual level data for everyone, but the loss function only wants to match the first moments between the `X0` mean and `X1` mean, so the individual level data for the treated units wasn't actually used. So the two functions `_torch` and `_retarget` are basically the same except in the first one, averaging over the treatment group is [implicit](https://github.com/apoorvalal/covariate_balancing_propensity_scores/blob/master/src/cbps_torch.py#L106) and in the second one, the averaging is explicit since the [loss function accepts the target averages directly](https://github.com/apoorvalal/covariate_balancing_propensity_scores/blob/master/src/cbps_retarget.py#L43). Check out the Wang/Zubizarreta paper linked in the readme, it is really good.

apoorvalal commented on src/cbps_torch.py:
```diff
@@ -0,0 +1,112 @@
+            out = np.exp(self.X_t @ self.theta)[self.w_t == 0]
+        elif self.estimand == "ATE":
+            out = np.exp(self.X @ self.theta)
+
```
meant to do this but now going to give you hw from 450b: add a diagnose() method that visualizes pre- and post-reweighting balance along covariates. Something like this https://cran.r-project.org/web/packages/cobalt/vignettes/love.plot.html

ivanhigueram committed:
Add dimension reduction via SVD

We add some docs to the class, maybe something more technical would be
better? Also, when X is big, might be convenient to reduce the space and
factorize X to its minimal expression using a low-rank approach. I am
hacking things here as I am picking the singular-value threshold myself,
but there are basic alternatives (np.linalg.matrix_rank?), although
inspired by LoRA, I wanted to try something more "aggressive".

Bugs/TODO:
----
 - Maybe self.X can be defined via @property decorator instead
 - A logger might be a good idea

ivanhigueram committed:
Clean notebook

apoorvalal commented on src/cbps_torch.py:
```diff
@@ -19,25 +19,65 @@ def __init__(
-        self.X = torch.from_numpy(X).float().to(device)
         self.W = torch.from_numpy(W).float().to(device)
 
+        # SVD for dimensionality reduction
```
add cite to https://www.jmlr.org/papers/volume19/17-777/17-777.pdf once you test this out. Have never thought low rank is all that useful in generic balancing problems, but makes sense for panel applications in particular. If you do this, allow tuning of k (`self.svd`).

apoorvalal commented:
LGTM. Thanks!
